internal/provider/services/gateway/gateway.go

```diff
@@ -0,0 +1,108 @@
+package gateway
+
+import (
+	"context"
+	"log"
+
+	"github.com/terrariumcloud/terrarium/internal/provider/services"
+	terrarium "github.com/terrariumcloud/terrarium/pkg/terrarium/provider"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	ConnectToVersionManagerError     = status.Error(codes.Unavailable, "Failed to connect to Version manager service.")
+	UnknownVersionManagerActionError = status.Error(codes.InvalidArgument, "Unknown Version manager action requested.")
+)
+
+type TerrariumGrpcGateway struct {
+	terrarium.UnimplementedPublisherServer
+	versionManagerClient services.VersionManagerClient
+}
+
+func New(versionManagerClient services.VersionManagerClient) *TerrariumGrpcGateway {
+	return &TerrariumGrpcGateway{
+		versionManagerClient: versionManagerClient,
+	}
+}
+
+// RegisterWithServer registers TerrariumGrpcGateway with grpc server
+func (gw *TerrariumGrpcGateway) RegisterWithServer(grpcServer grpc.ServiceRegistrar) error {
+	terrarium.RegisterPublisherServer(grpcServer, gw)
+	return nil
+}
+
+// Register new provider with Registrar service
+func (gw *TerrariumGrpcGateway) Register(ctx context.Context, request *terrarium.RegisterProviderRequest) (*terrarium.Response, error) {
+	return gw.RegisterWithClient(ctx, request, gw.versionManagerClient)
+}
+
+// RegisterWithClient calls Register on version manager client
+func (gw *TerrariumGrpcGateway) RegisterWithClient(ctx context.Context, request *terrarium.RegisterProviderRequest, client services.VersionManagerClient) (*terrarium.Response, error) {
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(
+		attribute.String("provider.name", request.GetName()),
+		attribute.String("provider.version", request.GetVersion()),
+	)
+
+	if res, delegateError := client.Register(ctx, request); delegateError != nil {
+		log.Printf("Failed: %v", delegateError)
+		span.RecordError(delegateError)
+		return nil, delegateError
+	} else {
+		log.Println("Done <= Registered")
+		span.AddEvent("Successful call to Register on Version Manager client.")
+		return res, nil
+	}
+}
+
+func (gw *TerrariumGrpcGateway) EndProvider(ctx context.Context, request *terrarium.EndProviderRequest) (*terrarium.Response, error) {
+	return gw.EndProviderWithClient(ctx, request, gw.versionManagerClient)
+}
+
+// EndVersionWithClient calls AbortProvider/AbortProviderVersion/PublishVersion on Version Manager client
+func (gw *TerrariumGrpcGateway) EndProviderWithClient(ctx context.Context, request *terrarium.EndProviderRequest, client services.VersionManagerClient) (*terrarium.Response, error) {
+
+	terminateProviderVersion := services.TerminateVersionRequest{
+		Provider: request.GetProvider(),
+	}
+
+	publishRequest := services.TerminateVersionRequest{
+		Provider: request.GetProvider(),
+	}
+
+	span := trace.SpanFromContext(ctx)
+
```

Comment:
You should set the span attributes here (like in the `RegisterWithClient` function above) for the request to this function.

Reply:
Added.